feat(ui-api, studies): optimize studies listing

smailio opened the pull request.

Commit:
feat(ui): add on click fetch and display list of non studies folder

Commit:
refactor(ui): fix build to pass npm lint add docs

Commit:
refactor(ui): make npm lint pass

Commit:
refactor(ui): make npm lint pass

Commit:
refactor(ui): move files, better naming

Commit:
refactor(ui): wip, add snack bar error alert for failed path load, refactor code into a functional style

Commit:
feat(ui): add workspace merge + optimize folder merge

Commit:
fix(ui): fix build

Commit:
fix(ui): fix build for real

Commit:
refactor(ui): refactor some test + ui

Commit:
refactor(ui): refactor some test + ui

Commit:
fix(ui): fix build

Commit:
refactor(ui): remove useless code, add doc, better naming

Commit:
fix(ui): fix tests

Commit:
feat(ui): handle click on root

Commit:
refactor(ui): add docs, renaming

Commit:
fix(ui): fis bug first level of folder wasn't initialized

Commit:
refactor(ui): make more modular

Commit:
refactor(ui): use i18n, make code more functionnal style

Commit:
feat(explorer): add parent path field, rename Dto object

Commit:
refactor(explorer): better doc string

Commit:
refactor(ui): adapt to new backend format

Commit:
fix(ui): over escaped special charachters

Commit:
fix(explorer): fix build

Commit:
fix(ui): fix build

Commit:
fix(ui): fix build

Commit:
feat(ui): make the scan non recursive by defaul, add an option for recursive scan, show only studies in the current folder by default

use a tree icon when we show studies for all descendants

Commit:
fix(ui): fix build npm lint

Commit:
fix(ui): fix build

Commit:
fix(ui): fix nonense tooltip messsage

Commit:
fix(ui): fix build

Commit:
fix(ui): fix build

Commit:
refactor(ui): better naming and messages

Commit:
refactor(ui): refactor duplicate code and fix style

Commit:
fix(ui): fix build

Commit:
refactor(ui): move code into clean sections

Commit:
refactor(ui): move test for study tree in their own folder

Commit:
refactor(ui): remove useless await

Commit:
Merge branch 'dev' into feature/615-optimize-listing-studies

Commit:
fix(build): add forgotten file

Commit:
Merge branch 'dev' into feature/615-optimize-listing-studies

Commit:
feat(ui): fetch current folder and not only children

Commit:
refactor(ui): add docs

Commit:
refactor(ui): small refactor

Commit:
fix(study-tree): fix default workspace not shown

Commit:
fix(study-tree): add docs

Commit:
fix(studies-ui): fetch only the current subfolders

Commit:
fix(studies-ui): don't fetch the subfolder under default workspace

Commit:
feature(studies): add has_children flag to non study folder

Commit:
fix(study-tree): wip

Commit:
fix(study-tree): wip

Commit:
fix(study-tree): refactor and add tests

Commit:
fix(study-tree): fix validation error

Commit:
fix(study-tree): fix stupid bug

Commit:
fix(study-tree): fix nmp lint

Commit:
fix(study-tree): remove console log, add locales

hdinia commented on webapp/public/locales/en/main.json:
```diff
@@ -642,7 +642,9 @@
   "studies.filters.strictfolder": "Show only direct folder children",
+  "studies.filters.showAllDescendants": "Show all children",
   "studies.scanFolder": "Scan folder",
+  "studies.requestDeepScan": "Recursive scan",
```
I suggest simplifying the i18n keys to match their UI labels more directly:

`studies.filters.showAllDescendants` → `studies.filters.showChildrens`
`studies.requestDeepScan` → `studies.recursiveScan`
(in the code you dont see the UI label so it can be a bit confusing when you search a specific key that correspond to the UI text)

hdinia commented on webapp/public/locales/en/main.json:
```diff
@@ -674,6 +676,10 @@
   "studies.variant": "Variant",
+  "studies.tree.error.failToFetchWorkspace": "Failed to load workspaces",
+  "studies.tree.error.failToFetchFolder": "Failed to load subfolders for {{path}}",
+  "studies.tree.error.detailsInConsole": "Details logged in the console",
```
For toast notification text, we should focus on user-friendly messages rather than technical details ("Details logged in the console"), I suggest you remove this one as you already raise an appropriate message for the error

smailio replied:
Those errors will be rarely met and when it happens someone will probably need to debug it, so that hint can be usefull no ?

hdinia replied:
having technical references in user-facing error messages breaks the consistency of the UI's tone and can confuse non-technical users, for devs who need to debug i think the console is already a standard place to look
of course you can keep the console logging for debugging 👍🏼

hdinia commented on webapp/src/components/App/Studies/StudiesList/index.tsx:
```diff
@@ -88,6 +90,7 @@ function StudiesList(props: StudiesListProps) {
   const [selectedStudies, setSelectedStudies] = useState<string[]>([]);
   const [selectionMode, setSelectionMode] = useState(false);
   const [confirmFolderScan, setConfirmFolderScan] = useState<boolean>(false);
+  const [isRecursiveScan, setIsRecursiveScan] = useState<boolean>(false);
```
Tip: TypeScript can infer the type from useState's default value, so you can simplify:
`useState<boolean>(false)` → `useState(false)`. Exception to this rule when initializing with a different type than what the state will eventually hold: 
```ts
// Starting empty but hold numbers
const [items, setItems] = useState<number[]>([]);
```

Commit:
Merge branch 'dev' into feature/615-optimize-listing-studies

hdinia commented on webapp/src/components/App/Studies/StudyTree/__test__/fixtures.ts:
```diff
@@ -0,0 +1,309 @@
+function mkStudyMetadata(folder: string, workspace: string): StudyMetadata {
+  return {
+    id: "xxxxxxxxxx",
+    name: "XXXXXXXXXX",
```
Should be something meaningful like:
id: "test-study-id",
name: "Test Study"

hdinia commented on webapp/src/components/App/Studies/StudyTree/__test__/fixtures.ts:
```diff
@@ -0,0 +1,309 @@
+      mkStudyMetadata("plop1/plop2/xxx1", "workspace"),
+      mkStudyMetadata("plop1/plop3/xxx2", "workspace"),
+      mkStudyMetadata("plop1/plop4/xxx3", "workspace"),
+      mkStudyMetadata("plouf1/plouf2/xxx4", "workspace2"),
```
please avoid using these kind of non-descriptive names, instead prefer names that have business meaning like:
"studies/projects/project1"
"studies/archived/study1"
"workspace1/active/study2"

hdinia commented on webapp/src/components/App/Studies/StudyTree/__test__/fixtures.ts:
```diff
@@ -0,0 +1,309 @@
+
+import { StudyMetadata, StudyType } from "@/common/types";
+
+function mkStudyMetadata(folder: string, workspace: string): StudyMetadata {
```
`mk` is an unclear abbreviation `mkStudyMetadata` -> `createStudyMetadata`

hdinia commented on webapp/src/components/App/Studies/StudyTree/__test__/utils.test.ts:
```diff
@@ -0,0 +1,125 @@
+        ],
+      };
+
+      const workspaces: string[] = ["a", "workspace1", "workspace2"];
```
the explicit type annotation is not necessary (inferred by TS)

hdinia commented on webapp/src/components/App/Studies/StudyTree/index.tsx:
```diff
@@ -0,0 +1,184 @@
+      // api doesn't allow to fetch the subfolders of the default workspace
+      return;
+    }
+    // Bug fix : this function used to take only the itemId and the selectedNode, and we used to initialize treeAfterWorkspacesUpdate
```
it's a good idea to document this kind of tricky bug 👍🏼

hdinia commented on webapp/src/components/App/Studies/StudyTree/index.tsx:
```diff
@@ -0,0 +1,184 @@
+      pathsToFetch = [`root${selectedNode.path}`];
+    }
+    const [treeAfterSubfoldersUpdate, failedPath] =
+      await fetchAndInsertSubfolders(pathsToFetch, treeAfterWorkspacesUpdate);
+    if (failedPath.length > 0) {
```
we usually always add a space before and after each code block

hdinia commented on webapp/src/services/api/study.ts:
```diff
@@ -48,6 +53,30 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getWorkspaces = async (): Promise<string[]> => {
```
return type not necessary here

smailio replied:
I keep it for documtentation purpose, that way the reader knows what the function returns at first glance

hdinia replied:
That's a fair point about documentation. However, there are a few reasons why I still recommend omitting the explicit return type here:

TS type inference provides the same information via hover/IDE tooling. The explicit type annotation is redundant, and maintaining explicit return types creates extra work etc.

If documentation is the main concern, consider adding a JSDoc 👍🏼

hdinia commented on webapp/src/services/api/study.ts:
```diff
@@ -48,6 +53,30 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
+export const getFolders = async (
+  workspace: string,
+  folderPath: string,
+): Promise<NonStudyFolderDTO[]> => {
```
you can omit the return type here

hdinia commented on webapp/src/services/api/study.ts:
```diff
@@ -434,8 +463,13 @@ export const updateStudyMetadata = async (
+export const scanFolder = async (
+  folderPath: string,
+  recursive = false,
+): Promise<void> => {
```
you can omit the return type here

hdinia commented on webapp/src/components/App/Studies/StudyTree/index.tsx:
```diff
@@ -0,0 +1,184 @@
+  // JSX
+  ////////////////////////////////////////////////////////////////
+
+  const buildTree = (children: StudyTreeNode[], parentId?: string) => {
```
`buildTree` is recreated on each render of StudyTree, which is potentially expensive for large tree structures with many children, also to keep the components cleaner it's better to refactor it into a dedicated module, here's what i suggest:

1. Extract the tree node rendering logic into a separate memoized component (add memo only if too many re-renders happens when opening/closing sub-folders)
2. Memoize the click handlers to prevent unnecessary re-renders (you can check if its really necessary)

Suggested implementation:

```typescript
const StudyTreeNode = memo(function StudyTreeNode({ node, parentId, onNodeClick }: StudyTreeNodeProps) {
  const id = parentId ? `${parentId}/${node.name}` : node.name;
  
  const handleClick = useCallback(() => {
    onNodeClick(id, node);
  }, [id, node, onNodeClick]);

  const isLoading = node.hasChildren && node.children.length === 0;

  return (
    <TreeItemEnhanced 
      itemId={id} 
      label={node.name} 
      onClick={handleClick}
    >
      {isLoading ? (
        <TreeItemEnhanced
          itemId={id + "loading"}
          label={t("studies.tree.fetchFolderLoading")}
        />
      ) : (
        node.children.map((child) => (
          <StudyTreeNode
            key={`${id}/${child.name}`}
            node={child}
            parentId={id}
            onNodeClick={onNodeClick}
          />
        ))
      )}
    </TreeItemEnhanced>
  );
});
```
```ts
// In StudyTree component:
function StudyTree() {
  const handleNodeClick = useCallback((itemId: string, node: StudyTreeNode) => {
    dispatch(updateStudyFilters({ folder: itemId }));
    updateTree(itemId, studiesTree, node);
  }, [dispatch, studiesTree]);

  return (
    <SimpleTreeView
      defaultExpandedItems={[...getParentPaths(folder), folder]}
      defaultSelectedItems={folder}
      sx={{ ... }}
    >
      <StudyTreeNode node={studiesTree} onNodeClick={handleNodeClick} />
    </SimpleTreeView>
  );
}
```

hdinia commented on webapp/src/components/App/Studies/StudyTree/index.tsx:
```diff
@@ -0,0 +1,184 @@
+        child.hasChildren && child.children.length === 0;
+      const id = parentId ? `${parentId}/${child.name}` : child.name;
+
+      if (shouldDisplayLoading) {
```
nothing wrong about this but generally prefer using prefix is, like "isDoingSomething" for flags (the common React boolean flag naming convention), here `isLoadingFolder` or `isLoading` seems better

MartinBelthle commented on antarest/study/model.py:
```diff
@@ -332,6 +332,22 @@ class NonStudyFolder(AntaresBaseModel):
     name: str
+    has_children: bool = Field(
+        alias="hasChildren",
+    )  # true when has non study folder children (false when has no children or only study children)
```
Not really clear. You could simplify by writing `# true when has at least one non-study-folder children`

MartinBelthle commented on antarest/study/storage/explorer_service.py:
```diff
@@ -33,18 +33,23 @@ def list_dir(
                 child_rel_path = child.relative_to(workspace.path)
-                directories.append(NonStudyFolder(path=child_rel_path, workspace=workspace_name, name=child.name))
+                has_children = has_non_study_folder(child)
+                directories.append(
```
Minor:
using the camel case to instantiate the class isn't perfect. You can try to use a dict and do NonStudyFolderDTO.model_validate(...) with the dict keys being has_children, it could work

Commit:
fix bugs add tests

Commit:
fix'(study-tree):  apply pr request

Commit:
fix(strudy-tree): fix build

Commit:
fix(study-tree): remove technical error message from snack bar

Commit:
fix(study-tree): remove useless comments

Commit:
fix(study-tree): remove return type

MartinBelthle reviewed (dismissed):
Okay for me for the back-end part

hdinia commented on webapp/public/locales/en/main.json:
```diff
@@ -675,6 +677,10 @@
   "studies.variant": "Variant",
+  "studies.tree.error.failToFetchWorkspace": "Failed to load workspaces",
+  "studies.tree.error.failToFetchFolder": "Failed to load subfolders for {{path}}",
+  "studies.tree.error.detailsInConsole": "Details logged in the console",
```
remove the `"studies.tree.error.detailsInConsole"` key as you dont use it anymore

hdinia commented on webapp/src/components/App/Studies/StudyTree/StudyTreeNode.tsx:
```diff
@@ -0,0 +1,61 @@
+        />
+      </TreeItemEnhanced>
+    );
+  }
```
space please 😄

hdinia reviewed (dismissed):
2 very minor comments, otherwise looks good to me, thanks for the changes made 👍🏼  don't forget to squash when merging to not keep the commit history

Commit:
Merge branch 'dev' into feature/615-optimize-listing-studies

Commit:
fix(study-tree): add function that was removed during merge

maugde committed:
chore(commands): fix user name display in commands (#2284)


A context variable is introduced to hold the use of the current request,
and of the current task.
Can be re-used in the future to avoid passing this context information
as an argument to all methods (for logging, etc).

Commit:
feature(explorer): add exceptions everywhere

Commit:
remove unused key

Commit:
remove try excepts

MartinBelthle committed:
feat(launcher): allow local launcher to work with xpress (#2251)


- allow user to launch a simulation with xpress and presolve in local 
- use simulator 8.8.11 instead of 8.8.5
- Don't import failed outputs inside the study
- separate logs in stdout and stderr as it was done with the slurm launcher
- Persist the logs if the simulation fails (by introducing a local_workspace)

sylvlecl committed:
feat(installer): update installer for v2.19 (#2297)

MartinBelthle commented on antarest/study/storage/utils.py:
```diff
@@ -495,3 +495,20 @@ def should_ignore_folder_for_scan(path: Path, filter_in: t.List[str], filter_out
+
+
+def is_non_study_folder(path: Path, filter_in: t.List[str], filter_out: t.List[str]) -> bool:
+    if not path.is_dir():
```
You can remove this check as you're already checking this inside the `is_study_folder` method

smailio replied:
The is_study_folder will return false if the path is not a dir. 

But the point is valid as this is checked in should_ignore_folder_for_scan. 

I did it like this on purpose, imo it's better for the reader to put this condition at the start, that way the reader get this case out of his mid and it's less mental load when they read what follows.

MartinBelthle commented on antarest/study/storage/explorer_service.py:
```diff
@@ -33,26 +33,37 @@ def list_dir(
         workspace = get_workspace_from_config(self.config, workspace_name, default_allowed=False)
         directory_path = get_folder_from_workspace(workspace, workspace_directory_path)
         directories = []
+        children = []
```
You can remove this

MartinBelthle commented on antarest/study/storage/explorer_service.py:
```diff
@@ -33,26 +33,37 @@ def list_dir(
+        children = []
+
         try:
+            # this block is skipper in case of permission error
```
typo: skipper -> skipped

MartinBelthle commented on antarest/study/storage/explorer_service.py:
```diff
@@ -33,26 +33,37 @@ def list_dir(
-                child_rel_path = child.relative_to(workspace.path)
-                directories.append(NonStudyFolder(path=child_rel_path, workspace=workspace_name, name=child.name))
+            for child in children:
+                # if we can't acess one child we skip it
```
typo: acess -> access

MartinBelthle reviewed (changes requested):
Minor changes

Commit:
fix(study-tree): fix typo

Commit:
fix(explorer): forgot to add file

Commit:
refactor(study-tree): add space

MartinBelthle committed:
feat(bc): add constraint duplication endpoint (#2295)

MartinBelthle commented on antarest/study/storage/utils.py:
```diff
@@ -495,3 +495,18 @@ def should_ignore_folder_for_scan(path: Path, filter_in: t.List[str], filter_out
+
+
+def has_non_study_folder(path: Path, filter_in: t.List[str], filter_out: t.List[str]) -> bool:
+    for sub_path in path.iterdir():
```
Could be simplified in 
`return any(is_non_study_folder(sub_path, filter_in, filter_out) for sub_path in path.iterdir())` no ?

Commit:
refactor(explorer): us any instead of for loop

Commit:
Merge branch 'dev' into feature/615-optimize-listing-studies
